### PR TITLE
improve index.js

### PR DESCRIPTION
**Pull request by fanatid**
Related with #1

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,9 +1,9 @@
...
-  var length = a.length
-  var buffer = new Buffer(length)
+module.exports = function reverse (src) {
+  var buffer = new Buffer(src.length)
 
-  for (var i = 0, j = length - 1; i < length; ++i, --j) {
-    buffer[i] = a[j]
+  for (var i = 0, j = src.length - 1; i <= j; ++i, --j) {
+    buffer[i] = src[j]
+    buffer[j] = src[i]
```
Comment (dcousens) [inline index.js:6]: Is this actually worth it?
I'd of though it'd create cache issues when iterating huge arrays (or even small ones?)

Comment (fanatid) [inline index.js:6]:
``` js
var Benchmark = require('benchmark')

var buffer = new Buffer(Math.pow(10, 6))

function reverseNew (src) {
  var buffer = new Buffer(src.length)

  for (var i = 0, j = src.length - 1; i <= j; ++i, --j) {
    buffer[i] = src[j]
    buffer[j] = src[i]
  }

  return buffer
}

function reverseOld (a) {
  var length = a.length
  var buffer = new Buffer(length)

  for (var i = 0, j = length - 1; i < length; ++i, --j) {
    buffer[i] = a[j]
  }

  return buffer
}

new Benchmark.Suite()
  .add('new', () => { reverseNew(buffer) })
  .add('old', () => { reverseOld(buffer) })
  .on('cycle', (event) => { console.log(String(event.target)) })
  .run()
```

results:

``` shell
// 10 bytes
new x 1,616,476 ops/sec ±1.16% (79 runs sampled)
old x 1,598,544 ops/sec ±0.93% (89 runs sampled)
// 1KB
new x 372,832 ops/sec ±1.60% (83 runs sampled)
old x 270,145 ops/sec ±1.87% (83 runs sampled)
// 1MB
new x 467 ops/sec ±2.97% (79 runs sampled)
old x 335 ops/sec ±1.51% (79 runs sampled)
```

node version: 5.7.1

Comment (dominictarr) [inline index.js:6]: it just occured to me that by using a temporary variable you could do an in-place reverse, which would skip the allocation. the only module that depends on this is bitcoinjs-lib, which needs to reverse hashes (so, mutable isn't a problem). However, that library seems to just use `[].reverse.call(buffer)` although it does depend on this module.

Comment (fanatid) [inline index.js:6]: `[].reverse.call(buffer)` too slow:
1MB:

``` shell
new x 461 ops/sec ±3.81% (74 runs sampled)
old x 334 ops/sec ±2.11% (78 runs sampled)
array x 11.74 ops/sec ±0.63% (33 runs sampled)
```

Comment (fanatid) [inline index.js:6]: @dominictarr inplace version located in file [inplace.js](https://github.com/crypto-browserify/buffer-reverse/blob/a4124fb8f46f94bbca4e41acc2124e0ac8bddb44/inplace.js)
@dcousens maybe make sense use inplace by default?

Comment (fanatid) [inline index.js:6]: https://github.com/bitcoinjs/bitcoinjs-lib/pull/557

**Comment by dcousens** [comment]
Cheers @fanatid :+1:

**Comment by fanatid** [comment]
@dcousens what about inplace by default?

**Comment by dcousens** [comment]
@fanatid less user friendly IMHO?

**Comment by fanatid** [comment]
`buffer.reverse` mutate data
`[].reverse.call` mutate data
why this package shouldn't do same by default?

if developer doesn't want mutated data in source buffer, he always make `reverse(new Buffer(buffer))` as with `buffer.reverse` or `[].reverse.call` ...

**Comment by dcousens** [comment]
@fanatid the summary of my arguments is essentially just:

> [Functional] Purity > Impurity

I don't think it is worth bumping major for.

**Comment by fanatid** [comment]
ok
please release minor version with speed improvements :)

**Comment by dcousens** [comment]
Published as `v1.0.1`

**Comment by dominictarr** [comment]
If you really want to maximizr low level performance you need to be able to
reason about memory allocations functional purity negates this, and in the
actual place this is used, the input is immediately thrown away.

Major bumps are cheap! And because it makes no difference to the dependant,
that only needs a patch.

**Pull request merged**